ml-agents/mlagents/trainers/trainer_controller.py

```diff
@@ -273,15 +287,12 @@ def advance(self, env: EnvManager) -> int:
                 if brain_name in self.trainer_metrics:
                     self.trainer_metrics[brain_name].add_delta_step(delta_time_step)
                 if step_info.has_actions_for_brain(brain_name):
-                    trainer.add_experiences(
+                    _processor = self.managers[brain_name].processor
+                    _processor.add_experiences(
                         step_info.previous_all_brain_info[brain_name],
                         step_info.current_all_brain_info[brain_name],
                         step_info.brain_name_to_action_info[brain_name].outputs,
                     )
-                    trainer.process_experiences(
```

Comment:
Should the trainer controller still have a reference to the trainers ?

Reply:
Good question - eventually the trainer controller will still need to call advance() on all the trainers for them to be able to query the queues. So, I think there will still be a need for the outer loop having some reference to the trainer.